homeassistant/components/sensor/modem_callerid.py

```diff
@@ -0,0 +1,125 @@
+"""
+A sensor to monitor incoming calls using a USB modem that supports caller ID.
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/sensor.modem_callerid/
+"""
+import logging
+import voluptuous as vol
+from homeassistant.const import (STATE_IDLE,
+                                 STATE_UNAVAILABLE,
+                                 EVENT_HOMEASSISTANT_STOP,
+                                 CONF_NAME,
+                                 CONF_DEVICE)
+from homeassistant.components.sensor import PLATFORM_SCHEMA
+from homeassistant.helpers.entity import Entity
+import homeassistant.helpers.config_validation as cv
+
+REQUIREMENTS = ['https://github.com/vroomfonde1/basicmodem'
+                '/archive/0.5.zip'
+                '#basicmodem==0.5']
+
+_LOGGER = logging.getLogger(__name__)
+DEFAULT_NAME = 'Modem CallerID'
+ICON = 'mdi:phone-clasic'
+DEFAULT_DEVICE = '/dev/ttyACM0'
+
+STATE_RING = 'ring'
+STATE_CALLERID = 'callerid'
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
+    vol.Optional(CONF_DEVICE, default=DEFAULT_DEVICE): cv.string
+})
+
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Setup modem caller id sensor platform."""
+    name = config.get(CONF_NAME)
+    port = config.get(CONF_DEVICE)
+    add_devices([ModemCalleridSensor(hass, name, port)])
+
+
+class ModemCalleridSensor(Entity):
+    """Implementation of USB modem callerid sensor."""
+
+    def __init__(self, hass, name, port):
+        """Initialize the sensor."""
+        from basicmodem.basicmodem import BasicModem as bm
+        self._state = STATE_UNAVAILABLE
+        self._attributes = {"cid_time": 0, "cid_number": '', "cid_name": ''}
+        self._name = name
+        self.port = port
+        self.modem = None
+
+        self.modem = bm(self.port, self._incomingcallcallback)
```

Comment:
Please move the modem initialization and state check into `setup_platform` so that you can avoid adding a device when the modem has not been initialized.